domain/controllerupgrader/service/service.go

```diff
@@ -0,0 +1,521 @@
+	if err != nil {
+		return errors.Capture(err)
+	}
+	return nil
+}
+
+// validateControllerCanBeUpgrade is a set of validation checks run on all
+// controller(s) in the cluster to make sure they are in a state suitable
+// for upgrading.
+//
+// Specifically as part of this validation we check that each controller is
+// running the expected version. If not it means that a previous upgrade is
+// still ongoing.
+//
+// The checks performed are not guaranteed to be valid after this func returns.
+//
+// The following errors may be expected:
+// - [controllerupgradererrors.ControllerUpgradeBlocker] describing a block that
+// exists preventing a controller upgrade from proceeding.
+func (s *Service) validateControllerCanBeUpgraded(
+	ctx context.Context,
+	currentVersion semversion.Number,
+) error {
+	controllerNodeVersions, err := s.ctrlSt.GetControllerNodeVersions(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"getting controller version for each node in cluster: %w", err,
+		)
+	}
+
+	// blockedNodes is the set of nodes that are blocking the controller
+	// upgrade.
+	blockedNodes := []string{}
+	for node, version := range controllerNodeVersions {
+		// if the current version for controllers is greater than that of this
+		// node, then this node still needs to be upgraded.
+		//
+		// It is on purpose that this is not an exact check. Due to the
+		// possability the model agent version could be out of sync with the
+		// controller version it is possible for the controller nodes to be
+		// running a version higher then the current version. This is ok and
+		// permissible.
+		if currentVersion.Compare(version) > 0 {
+			blockedNodes = append(blockedNodes, node)
+		}
+	}
+
+	if len(blockedNodes) > 0 {
+		return controllerupgradererrors.ControllerUpgradeBlocker{
+			Reason: fmt.Sprintf(
+				"controller nodes %v are not running controller version %q",
+				blockedNodes, currentVersion,
+			),
+		}
+	}
+
+	return nil
+}
+
+// validateControllerCanBeUpgradedTo checks that the controller can be upgraded
+// from the current version to the new desired version.
+//
+// The following errors may be expected:
+// - [controllerupgradererrors.DowngradeNotSupported] if the requested version
+// being upgraded to would result in a downgrade of the controller.
+// - [controllerupgradeerrors.VersionNotSupported] if the requested version
+// being upgraded to is more then a patch version upgrade.
+// - [controllerupgradererrors.ControllerUpgradeBlocker] describing a block that
+// exists preventing a controller upgrade from proceeding.
+func (s *Service) validateControllerCanBeUpgradedTo(
+	ctx context.Context,
+	desiredVersion semversion.Number,
+) error {
+	currentVersion, err := s.ctrlSt.GetControllerVersion(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"getting current controller version: %w", err,
+		)
+	}
+
+	// Check that a downgrade is not being attempted.
+	if currentVersion.Compare(desiredVersion) >= 0 {
```

Comment:
`desiredVersion.Compare(currentVersion) < 0` might be better.

Reply:
This is also a logic fix. Have fixed :)